### PR TITLE
New Job Opportunity also reflect on ColoredCow Website

### PR DESCRIPTION
**Pull request by am-Anmol**
Targets #794 

### Description
**Job opportunities will `create/update/delete` on the `ColoredCow website` also if updated from `Employee Portal`.**
- Changes made in `Modules/HR/Observers/Recruitment/JobObserver.php` :
     - **`Created` function (working):**
          - Check if `status` is `published` or not, if it is `not` then do nothing.
          - If it is `published` then creating a new entry of data in `wp_post` table with `post_title,post_content,post_type` and a new entry in `wp_postmeta` table with `hr_id` and value of `hr_id` from employee portal `hr_jobs` table.
     - **`Updated` function (working):**
          - Check if `status` is `published` or not, if it is `not` then do nothing.
          - If it is `published` then creating a new entry of data in `wp_post` table with `post_title,post_content,post_type` and a new entry in `wp_postmeta` table with `hr_id` and value of `hr_id` from employee portal `hr_jobs` table.
          - Then finding `post id` by matching `hr_id` in `wp_postmeta` table (Website) with `hr_id` in `hr_jobs` table (employee portal).
          - After that finding an old post with `post id` and deleting the old post.
     - **`Deleted` function (working):**
          - Matching `post_title` and `post_type` with records and deleting the post.
      
- **Prerequisites Required:**
     - Update `config/database.php` with the correct database configurations.
     - image

**References:**
Check out the Corcel package to easily create entries in the WordPress database from Laravel
https://github.com/corcel/corcel
### Checklist:

- [x] I have performed a self-review of my own code.

### CODE REVIEW

File: Modules/HR/Observers/Recruitment/JobObserver.php [hunk 1]
```diff
@@ -16,5 +17,49 @@ class JobObserver
     public function created(Job $job)
     {
         $job->rounds()->attach(Round::all()->pluck('id')->toArray());
+        $data = request()->all();
+
+        if ($data['status'] == 'published') {
+            $Corcel = new Corcel();
+            $Corcel->post_title = $data['title'];
+            $Corcel->post_content = $data['description'];
+            $Corcel->post_type = 'career';
```
Comment (GauravGusain98) [inline Modules/HR/Observers/Recruitment/JobObserver.php:26]: Please use the value from config files instead of hardcoding. This way if we want to change the value of post_type, we have to only change that in one place instead of changing at every place.

Comment (GauravGusain98) [inline Modules/HR/Observers/Recruitment/JobObserver.php:26]: Please use the value from config files instead of hardcoding. This way if we want to change the value of post_type, we have to only change that in one place instead of changing at every place.

File: Modules/HR/Observers/Recruitment/JobObserver.php [hunk 2]
```diff
@@ -16,5 +17,49 @@ class JobObserver
...
+            $Corcel->save();
+            $Corcel->saveMeta('hr_id', $job['id']);
+            $post = Corcel::hasMeta('hr_id', $job['id'])->first();
+            $corcel = Corcel::find($post->ID);
+            $corcel->delete();
```
Comment (GauravGusain98) [inline Modules/HR/Observers/Recruitment/JobObserver.php:51]: This is not looking correct to me as we need to update the record here. But in this case, we are creating a new entry and deleting the old one. Please use the update approach for this.

**Review by GauravGusain98 — CHANGES_REQUESTED** [review]
Please make the suggested changes.

File: Modules/HR/Observers/Recruitment/JobObserver.php (hunk 2 above)
Comment (am-Anmol) [reply Modules/HR/Observers/Recruitment/JobObserver.php:51]: OK, I'll update this approach.

File: Modules/HR/Observers/Recruitment/JobObserver.php (hunk 1 above)
Comment (am-Anmol) [reply Modules/HR/Observers/Recruitment/JobObserver.php:26]: Ok, I make these changes.

**Comment by am-Anmol** [comment]
**Update function approach:**
- Check if `status` is `published` or not, if it is `not` then do nothing.
- If it is `published` then find `post ID` in `wp_post` table using `job_id`.
- Then finding `post` with `post ID` and updating respective columns.

File: Modules/HR/Observers/Recruitment/JobObserver.php [hunk 3]
```diff
@@ -16,5 +17,42 @@ class JobObserver
...
+     * Listen to the Job update event.
+     *
+     * @param  \Modules\HR\Entities\Job  $job
+     * @return void
+     */
+    public function updated(Job $job)
+    {
+        $data = request()->all();
+
+        if ($data['status'] == 'published') {
```
Comment (GauravGusain98) [inline Modules/HR/Observers/Recruitment/JobObserver.php:42]: Just a doubt, can there be a case when the status is not published and we can update the job? If yes, then we need to update the job details for both the case.

**Review by GauravGusain98 — COMMENTED** [review]
Please check the comment

**Comment by GauravGusain98** [comment]
Please check the above comment

File: Modules/HR/Observers/Recruitment/JobObserver.php (hunk 3 above)
Comment (am-Anmol) [reply Modules/HR/Observers/Recruitment/JobObserver.php:42]: Yes, this can be a possible scenario then we have to update both of them.
If we update the job from the portal and mark status to published it also gets updated on a website post.

Comment (GauravGusain98) [reply Modules/HR/Observers/Recruitment/JobObserver.php:42]: Let me rephrase my question, 
I was saying that suppose a job is not published but you updated the job details in the portal. In this case as well, the job details should get updated in the website post. But as per the code, it looks like that the job will only get updated when the status is published.

Also what about the case when we update the job from `published` to `draft`. I don't think in that case it will mark the job as a draft on the website. 

Let me know if there is any doubt in understanding.

**Review by GauravGusain98 — CHANGES_REQUESTED** [review]
Please check the comments and if required do the requested changes.

File: Modules/HR/Observers/Recruitment/JobObserver.php [hunk 4]
```diff
@@ -16,5 +17,42 @@ class JobObserver
...
+            $Corcel = new Corcel();
+            $Corcel->post_title = $data['title'];
+            $Corcel->post_content = $data['description'];
+            $Corcel->post_type = config('hr.post-type.career');
+            $Corcel->save();
+            $Corcel->saveMeta('hr_id', $job['id']);
+        }
+    }
```
Comment (GauravGusain98) [inline Modules/HR/Observers/Recruitment/JobObserver.php:30]: Let me rephrase my question, 
I was saying that suppose a job is not published but you updated the job details in the portal. In this case as well, the job details should get updated in the website post. But as per the code, it looks like that the job will only get updated when the status is published.

Also what about the case when we update the job from `published` to `draft`. I don't think in that case it will mark the job as a draft on the website. 

Let me know if there is any doubt in understanding.

**Review by GauravGusain98 — CHANGES_REQUESTED** [review]
Please do the following changes.

**Comment by am-Anmol** [comment]
**Changes in created and updated function:**
- Created:
    - Checking if the status is `not published` then saving the post as a draft.
- Updated:
   - Checking if the status is `published` then updating the post as a `publish`, otherwise updating as `draft` only.

File: Modules/HR/Observers/Recruitment/JobObserver.php (hunk 4 above)
Comment (am-Anmol) [reply Modules/HR/Observers/Recruitment/JobObserver.php]: Yeah, I have understood your scenario and updated the code as per all scenarios possible while creating and updating any post through an employee portal.

File: Modules/HR/Observers/Recruitment/JobObserver.php [hunk 5]
```diff
@@ -16,5 +17,50 @@ class JobObserver
...
+        $post = Corcel::hasMeta('hr_id', $job['id'])->first();
+        $Corcel = Corcel::find($post->ID);
+        $Corcel->post_title = $data['title'];
+        $Corcel->post_content = $data['description'];
+        $Corcel->post_type = config('hr.post-type.career');
+        if ($data['status'] == 'published') {
+            $Corcel->post_status = 'publish';
+        } else {
+            $Corcel->post_status = 'draft';
+        }
```
Comment (GauravGusain98) [inline Modules/HR/Observers/Recruitment/JobObserver.php:52]:
```suggestion
      $Corcel->post_status = $data['status'] == 'published' ? 'published' : 'draft';
```

**Review by GauravGusain98 — CHANGES_REQUESTED** [review]
Please check the suggestion

File: Modules/HR/Observers/Recruitment/JobObserver.php (hunk 5 above)
Comment (am-Anmol) [reply Modules/HR/Observers/Recruitment/JobObserver.php]: Ok I make changes

**Review by GauravGusain98 — APPROVED** [review]
Looks good to me
